dubbo-plugin/dubbo-qos/src/main/java/com/alibaba/dubbo/qos/protocol/QosProtocolWrapper.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alibaba.dubbo.qos.protocol;
+
+import com.alibaba.dubbo.common.Constants;
+import com.alibaba.dubbo.common.URL;
+import com.alibaba.dubbo.qos.server.Server;
+import com.alibaba.dubbo.rpc.Exporter;
+import com.alibaba.dubbo.rpc.Invoker;
+import com.alibaba.dubbo.rpc.Protocol;
+import com.alibaba.dubbo.rpc.RpcException;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.alibaba.dubbo.common.Constants.ACCEPT_FOREIGN_IP;
+import static com.alibaba.dubbo.common.Constants.QOS_ENABLE;
+import static com.alibaba.dubbo.common.Constants.QOS_PORT;
+
+public class QosProtocolWrapper implements Protocol {
+    private static AtomicBoolean hasStarted = new AtomicBoolean(false);
+
+    private Protocol protocol;
+
+    public QosProtocolWrapper(Protocol protocol) {
+        if (protocol == null) {
+            throw new IllegalArgumentException("protocol == null");
+        }
+        this.protocol = protocol;
+    }
+
+    @Override
+    public int getDefaultPort() {
+        return protocol.getDefaultPort();
+    }
+
+    @Override
+    public <T> Exporter<T> export(Invoker<T> invoker) throws RpcException {
+        if (Constants.REGISTRY_PROTOCOL.equals(invoker.getUrl().getProtocol())) {
+            startQosServer(invoker.getUrl());
+            return protocol.export(invoker);
+        }
+        return protocol.export(invoker);
+    }
+
+    @Override
+    public <T> Invoker<T> refer(Class<T> type, URL url) throws RpcException {
+        if (Constants.REGISTRY_PROTOCOL.equals(url.getProtocol())) {
+            startQosServer(url);
+            return protocol.refer(type, url);
+        }
+        return protocol.refer(type, url);
+    }
+
+    @Override
+    public void destroy() {
+        protocol.destroy();
+    }
+
+    private void startQosServer(URL url) {
+        if (!hasStarted.compareAndSet(false, true)) {
+            return;
+        }
+
+        try {
+            boolean qosEnable = Boolean.parseBoolean(url.getParameter(QOS_ENABLE,"true"));
```

Comment:
Why is this option default to true? Now if we start a dubbo client and a dubbo server in one machine, there is always an error message occurred...

ERROR [com.alibaba.dubbo.qos.server.Server]  [DUBBO] qos-server can not bind localhost:22222, dubbo version: 2.0.1, current host:
java.net.BindException: Address already in use